File: src-tauri/src/app/commands/upgrade.rs
use crate::app::modules::{
    core::{install::Install, upgrade::Upgrade},
    io::use_io,
};
use clap::Args;

#[derive(Args)]
pub struct UpgradeCommand {
    /// Mods to upgrade (id, slug or substring). If empty, upgrades all.
    pub mods: Vec<String>,

    #[arg(long)]
    pub no_cache: bool,

    #[arg(long)]
    pub force_rehash: bool,
}

impl UpgradeCommand {
    pub async fn handle(&self) {
        let io = use_io();

        match Upgrade::run(&self.mods).await {
            Ok(result) => {
                if result.upgraded.is_empty() {
                    io.info("All selected mods are already up to date");
                    return;
                }

                for (key, before, after) in result.upgraded {
                    io.success(&format!(
                        "Upgraded {}: {} → {}",
                        key,
                        before.unwrap_or_else(|| "-".into()),
                        after.unwrap_or_else(|| "-".into())
                    ));
                }

                if result.unchanged > 0 {
                    io.info(&format!(
                        "{} mod(s) were already up to date",
                        result.unchanged
                    ));
                }

                if let Err(e) = Install::run(self.no_cache, self.force_rehash).await {
                    io.error(&format!("Installation failed after upgrade: {}", e), None);
                    return;
                }

                io.success("All upgraded mods installed successfully.");
            }
            Err(msg) => io.error(&msg, None),
        }
    }
}

File: packages/cli/src/commands/upgrade.ts
import type {Command} from "commander";
import {Install, type ModManager, Upgrade} from "@mcpm/core";

export function registerUpgrade(program: Command, getManager: () => Promise<ModManager>): void {
  program
    .command("upgrade")
    .description("Upgrade mods, datapacks, resourcepacks and shaderpacks to newer compatible versions")
    .argument("[mods...]", "Items to upgrade (id, slug or substring)")
    .option("--no-cache", "Skip cache")
    .option("--force-rehash", "Force re-verification of hashes")
    .option("--ignore-constraints", "Ignore version constraints")
    .option(
      "--disable-unresolved",
      "Disable mods that have no compatible version for the target Minecraft version instead of failing",
    )
    .option(
      "--enable-resolved",
      "Re-enable disabled mods that now have a compatible version for the target Minecraft version",
    )
    .action(async (mods: string[], opts) => {
      const manager = await getManager();
      const io = manager.io;
      try {
        const result = await Upgrade.runWithManager(
          manager,
          mods,
          opts.ignoreConstraints === true,
          opts.disableUnresolved === true,
          opts.enableResolved === true,
        );

        for (const key of result.disabled) {
          io.warn(
            `Disabled ${key}: no compatible version found for Minecraft ${manager.manifestService.manifest.minecraft_version}`,
          );
        }

        for (const key of result.enabled) {
          io.success(`Enabled ${key}: a compatible version was found`);
        }

        for (const key of result.stillUnresolved) {
          io.info(
            `${key} is disabled and still has no compatible version for Minecraft ${manager.manifestService.manifest.minecraft_version}`,
          );
        }

        if (
          result.upgraded.length === 0 &&
          result.disabled.length === 0 &&
          result.stillUnresolved.length === 0 &&
          result.enabled.length === 0
        ) {
          io.info("All selected mods are already up to date");
          return;
        }

        for (const [key, before, after] of result.upgraded) {
          io.success(`Upgraded ${key}: ${before ?? "-"} → ${after ?? "-"}`);
        }

        if (result.unchanged > 0) {
          io.info(`${result.unchanged} mod(s) were already up to date`);
        }

        await Install.runWithManager(manager, opts.noCache !== true, opts.forceRehash === true);
        io.success("All upgraded mods installed successfully.");
      } catch (e) {
        io.error(e instanceof Error ? e.message : String(e));
        process.exit(1);
      }
    });
}
